### PR TITLE
fix(build): implement plugin system for app distribution

### PR DESCRIPTION
**Pull request by tido64**
### Description

Implemented plugin system for app distribution.

### Test plan

The test plan is similar to #1739. I've updated the `README`, but let me know if something is unclear or missing.

```sh
yarn
cd incubator/build
yarn build --dependencies

# Local deployment should still work
yarn rnx-build --platform android ../../packages/test-app

# Firebase config should be picked up if targeting device
yarn rnx-build --platform android --device-type device ../../packages/test-app

# Local deployment to devices should still work
yarn rnx-build --platform android --deploy local-only --device-type device ../../packages/test-app
```

### CODE REVIEW

File: incubator/build-plugin-firebase/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    "url": "https://github.com/microsoft/rnx-kit",
+    "directory": "incubator/build-plugin-firebase"
+  },
+  "scripts": {
+    "build": "rnx-kit-scripts build",
+    "format": "rnx-kit-scripts format",
+    "lint": "rnx-kit-scripts lint"
+  },
+  "peerDependencies": {
+    "@rnx-kit/build": "*"
```
Comment (afoxman) [inline incubator/build-plugin-firebase/package.json:28]: This should point to a specific version range of the package.

File: incubator/build/src/distribution.ts [hunk 2]
```diff
@@ -1,37 +1,33 @@
...
-  if (deployment === "autodetect") {
-    switch (service) {
-      case "firebase": {
-        return import("./distribution/firebase");
-      }
-    }
+  [plugin, pluginConfig]: Partial<Plugin>,
+  projectRoot: string
+): Promise<DistributionPlugin> {
+  if (deployment === "auto" && typeof plugin === "string") {
```
Comment (afoxman) [inline incubator/build/src/distribution.ts:17]: "auto" doesn't seem to be tied to any decision making logic. When the second plugin comes along, how will auto pick between them?

Maybe it's more of a choice between local and device?

I may be missing something obvious.

**Review by afoxman — APPROVED** [review]

File: incubator/build/src/distribution.ts (hunk 2 above)
Comment (tido64) [reply incubator/build/src/distribution.ts:17]: The "full" list of values would probably be `auto`, `local-only`, `remote-only`. `auto` here basically means try remote first, then fallback to local if no plugin is specified. You can only specify one distribution plugin.

Comment (tido64) [reply incubator/build/src/distribution.ts]: I renamed the values. Hopefully this is less confusing:

- `auto` -> `remote-first`
- `local` -> `local-only`

**Pull request merged**